### ISSUE TITLE
Uplift unvalidated logic to `ZeroMap`

### ISSUE THREAD

**Issue by robertbastian** [unicode-org/icu4x#5660]
Having unvalidated types for `ZeroMap`s is annoying. We have `UnavalidatedTinyAsciiStr`, but for `struct TimeZoneBcp47Id(TinyAsciiStr<8>)` we additionally need `struct UnvalidatedTimeZoneBcp47Id(UnvalidatedTinyAsciiStr<8>)`.

It might be better to have an `UnvaliatedZeroMap` that doesn't have the iteration methods, and which just skips the key validation internally.

**Comment by Manishearth** [comment]
I also think perhaps we should have `Unvalidated<T>` that wraps ULE types? The Ord impl could be byte-ord. It would be nice to have `ZeroVec<Unvalidated<something>>`

An UnvalidatedZeroMap isn't a bad idea either.

**Comment by sffc** [comment]
>for `struct TimeZoneBcp47Id(TinyAsciiStr<8>)` we additionally need `struct UnvalidatedTimeZoneBcp47Id(UnvalidatedTinyAsciiStr<8>)`.

No. Pure wrapper types that do not enforce additional invariants should not have their own ULE types. We discussed this not too long ago (can dig up the notes if needed)

**Comment by sffc** [comment]
Basically, I don't think there is anything wrong with the current model. I think it is quite clean.

There are just a few unvalidated types at the root:

- `[u8]` for variable-length things
- `RawBytesULE<N>` for fixed-length things
- `potential_utf::PotentialUtf8` for variable-length things that want to be rendered as strings in human-readable mode
- `tinystr::UnvalidatedTinyAsciiStr`, which should probably be `potential_utf::PotentialFixedLengthUtf8`, for fixed-length things that want to be rendered as strings in human-readable mode

As a table:

| | Fixed-Length | Variable-Length |
|---|---|---|
| Stringy | UnvalidatedTinyAsciiStr | PotentialUtf8 |
| Not Stringy | RawBytesULE | [u8] |

There should never be `UnvalidatedTimeZoneBcp47Id`. You should just use `UnvalidatedTinyAsciiStr`.

**Comment by Martin005** [comment]
@robertbastian It looks like this issue may be resolved (or at least no longer relevant):
- `TimeZoneBcp47Id(pub TinyAsciiStr<8>)` was renamed to `TimeZone(pub TinyAsciiStr<8>)` in [#6123](https://github.com/unicode-org/icu4x/pull/6123/changes#diff-fb3785cedb3e013c8bbb48659756a29ca6bd4edd4394ed1bcfebdc332d68d8c1R74). Then in [#6460](https://github.com/unicode-org/icu4x/pull/6460/changes#diff-dd9bce37cba94e980431a686c194069fb7d4ed0275434b9301a369ab4e44558bR181) the inner type was changed to `Subtag`. The current `TimeZone(pub Subtag)` wraps a validated `Subtag`, so there's no longer a need for an `UnvalidatedTimeZoneBcp47Id`.

- The existing pattern of using unvalidated wrapper types as ZeroMap keys already covers the use cases — `UnvalidatedTinyAsciiStr<N>` for fixed-length stringy keys, `PotentialUtf8` for variable-length stringy keys. These are used throughout (e.g. `ZeroMap<'data, UnvalidatedTinyAsciiStr<2>, Region>` in the locale provider).

Should this issue be closed, or is there still a specific use case that isn't well-served by the current approach?
